File: androidApp/src/main/kotlin/se/soderbjorn/termtastic/android/ui/ReformatIcon.kt
/**
 * The "reformat" toolbar glyph for the Termtastic Android app.
 *
 * Draws the same icon the Electron/web client uses for its reformat-on-resize
 * action: a terminal rectangle with a chevron tucked against each side wall,
 * reading as "reflow the contents to fit the width". This replaces the earlier
 * circular refresh icon, which looked like a reload, so the reformat affordance
 * is visually consistent across the web, desktop and mobile clients.
 *
 * @see se.soderbjorn.termtastic.android.ui.TerminalScreen
 */
package se.soderbjorn.termtastic.android.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp

/**
 * Renders the reformat glyph, stroked in [tint], inside an icon-sized box.
 *
 * Called from [TerminalScreen]'s top-bar reformat [androidx.compose.material3.IconButton];
 * mirrors the web client's 24×24 SVG (a `rect` plus two inward chevrons) so the
 * stroke geometry matches pixel-for-pixel after scaling to the available size.
 *
 * @param tint stroke colour for the rectangle and chevrons (theme accent).
 * @param modifier layout modifier; sized to 24dp by default to match Material
 *   [androidx.compose.material3.Icon] dimensions.
 */
@Composable
fun ReformatIcon(
    tint: Color,
    modifier: Modifier = Modifier,
) {
    Canvas(
        modifier = modifier
            .size(24.dp)
            .semantics { contentDescription = "Reformat" },
    ) {
        // The source artwork is authored in a 24×24 viewport; derive a uniform
        // scale so the icon stays crisp at whatever size the layout grants.
        val s = size.minDimension / 24f
        val stroke = Stroke(
            width = 2f * s,
            cap = StrokeCap.Round,
            join = StrokeJoin.Round,
        )
        // Terminal frame: <rect x="3" y="5" width="18" height="14" rx="1.5"/>.
        drawRoundRect(
            color = tint,
            topLeft = Offset(3f * s, 5f * s),
            size = Size(18f * s, 14f * s),
            cornerRadius = CornerRadius(1.5f * s, 1.5f * s),
            style = stroke,
        )
        // Left chevron: <polyline points="7 10 4 12 7 14"/>.
        drawPath(
            path = Path().apply {
                moveTo(7f * s, 10f * s)
                lineTo(4f * s, 12f * s)
                lineTo(7f * s, 14f * s)
            },
            color = tint,
            style = stroke,
        )
        // Right chevron: <polyline points="17 10 20 12 17 14"/>.
        drawPath(
            path = Path().apply {
                moveTo(17f * s, 10f * s)
                lineTo(20f * s, 12f * s)
                lineTo(17f * s, 14f * s)
            },
            color = tint,
            style = stroke,
        )
    }
}

File: iosApp/iosApp/Views/ReformatIcon.swift
// ReformatIcon.swift
//
// The "reformat" toolbar glyph for the Termtastic iOS app.
//
// Draws the same icon the Electron/web client uses for its reformat-on-resize
// action: a terminal rectangle with a chevron tucked against each side wall,
// reading as "reflow the contents to fit the width". This replaces the earlier
// `arrow.clockwise` SF Symbol, which looked like a reload, so the reformat
// affordance is visually consistent across the web, desktop and mobile clients.

import SwiftUI

/// The reformat glyph as a `Shape`, authored in a 24×24 viewport and scaled to
/// fit whatever frame it is given. Mirrors the web client's SVG (`rect` plus two
/// inward chevrons) so the stroke geometry matches after scaling.
///
/// - SeeAlso: `ReformatIcon`
struct ReformatGlyph: Shape {
    func path(in rect: CGRect) -> Path {
        let s = min(rect.width, rect.height) / 24
        func p(_ x: CGFloat, _ y: CGFloat) -> CGPoint {
            CGPoint(x: rect.minX + x * s, y: rect.minY + y * s)
        }

        var path = Path()
        // Terminal frame: <rect x="3" y="5" width="18" height="14" rx="1.5"/>.
        path.addRoundedRect(
            in: CGRect(x: rect.minX + 3 * s, y: rect.minY + 5 * s,
                       width: 18 * s, height: 14 * s),
            cornerSize: CGSize(width: 1.5 * s, height: 1.5 * s)
        )
        // Left chevron: <polyline points="7 10 4 12 7 14"/>.
        path.move(to: p(7, 10))
        path.addLine(to: p(4, 12))
        path.addLine(to: p(7, 14))
        // Right chevron: <polyline points="17 10 20 12 17 14"/>.
        path.move(to: p(17, 10))
        path.addLine(to: p(20, 12))
        path.addLine(to: p(17, 14))
        return path
    }
}

/// Renders the reformat glyph at toolbar-icon size, stroked in the current
/// foreground style. Used by `TerminalScreen`'s top-bar reformat button;
/// callers apply `.foregroundStyle(Palette.headerAccent)` to tint it.
struct ReformatIcon: View {
    var body: some View {
        ReformatGlyph()
            .stroke(style: StrokeStyle(lineWidth: 1.8, lineCap: .round, lineJoin: .round))
            .frame(width: 22, height: 22)
            .accessibilityLabel("Reformat")
    }
}
